source/internal/interactive_batch.cpp

```diff
@@ -0,0 +1,275 @@
+#include "interactive_batch.h"
+#include "interactive_session.h"
+#include "common.h"
+
+#include <functional>
+
+namespace mixer_internal
+{
+
+interactive_batch_entry_internal::interactive_batch_entry_internal()
+	: next(nullptr), value(rapidjson::kObjectType)
+{
+}
+
+interactive_batch_array_internal::interactive_batch_array_internal()
+	: value(rapidjson::kArrayType)
+{
+}
+
+int interactive_batch_begin(interactive_session session, interactive_batch* batchPtr, char *methodName, on_get_batch_params getParams)
+{
+	if (nullptr == session || nullptr == batchPtr || nullptr == methodName || nullptr == getParams)
+	{
+		return MIXER_ERROR_INVALID_POINTER;
+	}
+
+	std::auto_ptr<interactive_batch_internal> batchInternal(new interactive_batch_internal());
+	batchInternal->session = reinterpret_cast<interactive_session_internal*>(session);
+	batchInternal->priority = RPC_PRIORITY_DEFAULT;
+	batchInternal->method = methodName;
+	batchInternal->getParams = getParams;
+	batchInternal->firstEntry = nullptr;
+	batchInternal->lastEntry = nullptr;
+	batchInternal->document = std::make_shared<rapidjson::Document>();
+	*batchPtr = reinterpret_cast<interactive_batch>(batchInternal.release());
+
+	return MIXER_OK;
+}
+
+int interactive_batch_add_entry(interactive_batch batch, interactive_batch_entry* entry)
+{
+	if (nullptr == batch || nullptr == entry)
+	{
+		return MIXER_ERROR_INVALID_POINTER;
+	}
+
+	interactive_batch_internal* batchInternal = reinterpret_cast<interactive_batch_internal*>(batch);
+	interactive_batch_entry_internal* entryInternal = new interactive_batch_entry_internal();
+	if (nullptr == batchInternal->firstEntry)
+	{
+		batchInternal->firstEntry = entryInternal;
+		batchInternal->lastEntry = entryInternal;
+	}
+	else
+	{
+		batchInternal->lastEntry->next = entryInternal;
+		batchInternal->lastEntry = entryInternal;
+	}
+
+	*entry = reinterpret_cast<interactive_batch_entry>(entryInternal);
+
+	return MIXER_OK;
+}
+
+int interactive_batch_iterate_entries(interactive_batch_internal* batch, on_get_entry_params getParams)
+{
+	if (nullptr == getParams)
+	{
+		return MIXER_ERROR_INVALID_POINTER;
+	}
+
+	interactive_batch_entry_internal* freePtr = nullptr;
+	interactive_batch_entry_internal* next = batch->firstEntry;
+	while (nullptr != next) {
+		getParams(next);
+		freePtr = next;
+		next = next->next;
+		delete freePtr;
+	}
+
+	batch->firstEntry = nullptr;
+	batch->lastEntry = nullptr;
+
+	return MIXER_OK;
+}
+
+int interactive_batch_free(interactive_batch_internal* batch)
+{
+	if (nullptr == batch)
+	{
+		return MIXER_ERROR_INVALID_POINTER;
+	}
+
+	interactive_batch_entry_internal* freePtr = nullptr;
+	interactive_batch_entry_internal* next = batch->firstEntry;
+	while (nullptr != next) {
+		freePtr = next;
+		next = next->next;
+		delete freePtr;
+	}
```

Comment:
If there's a failure between `interactive_batch_begin` and `interactive_batch_commit`, will all created entries leak?

Reply:
Good point. I'll make `interactive_batch_free` public so you can call it in an error state. Would then also want to make `_commit` not free the object so you consistently call `_free` every time.

That said, what do you think about having `interactive_batch_commit` free the entries (so you don't iterate the linked list again) but not the batch op object? This detail shouldn't be exposed to the user at all.

Reply:
I think we can get away with not having the caller own anything but `interactive_batch_op` which would greatly simplify memory management. See my comment below and let's chat.